File: mobile/ios/VetCare/CustomerModels.swift
//
//  CustomerModels.swift
//  VetCare
//
//  Codable models for pet-owner authentication and pet data.
//

import Foundation

// MARK: - Auth

struct CustomerLoginRequest: Encodable {
    let identifier: String
    let password: String
}

struct CustomerLoginResponse: Decodable {
    let data: CustomerLoginData
}

struct CustomerLoginData: Decodable {
    let customer: Customer
    let token: String
}

// MARK: - Customer

struct Customer: Codable, Identifiable {
    let customerId: String
    let firstName: String
    let lastName: String
    let email: String?
    let phone: String
    let alternatePhone: String?
    let address: String?
    let city: String?
    let nic: String?
    let preferredContactMethod: String?
    let emergencyContact: String?
    let emergencyPhone: String?
    var passwordMustChange: Bool
    let createdAt: String?

    var id: String { customerId }
    var fullName: String { "\(firstName) \(lastName)" }
    var initials: String { "\(firstName.prefix(1))\(lastName.prefix(1))" }

    var memberSince: String? {
        guard let raw = createdAt else { return nil }
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let date = iso.date(from: raw) ?? ISO8601DateFormatter().date(from: raw)
        guard let d = date else { return nil }
        return "\(Calendar.current.component(.year, from: d))"
    }

    var preferredContactDisplay: String {
        switch preferredContactMethod {
        case "phone": return "Phone"
        case "email": return "Email"
        case "sms":   return "SMS"
        default:      return "—"
        }
    }

    enum CodingKeys: String, CodingKey {
        case customerId           = "customer_id"
        case firstName            = "first_name"
        case lastName             = "last_name"
        case email
        case phone
        case alternatePhone       = "alternate_phone"
        case address
        case city
        case nic
        case preferredContactMethod = "preferred_contact_method"
        case emergencyContact     = "emergency_contact"
        case emergencyPhone       = "emergency_phone"
        case passwordMustChange   = "password_must_change"
        case createdAt            = "created_at"
    }
}

// MARK: - Pets

struct PetsResponse: Decodable {
    let data: PetsData
}

struct PetsData: Decodable {
    let pets: [Pet]
}

struct Pet: Codable, Identifiable, Hashable {
    let petId: String
    let name: String
    let species: String
    let breed: String?
    let gender: String?
    let dateOfBirth: String?
    let color: String?
    // node-postgres serialises DECIMAL columns as strings to preserve precision
    let weightCurrent: String?
    let isNeutered: Bool
    let allergies: String?
    let specialNeeds: String?

    var id: String { petId }

    enum CodingKeys: String, CodingKey {
        case petId = "pet_id"
        case name = "pet_name"
        case species
        case breed
        case gender
        case dateOfBirth = "date_of_birth"
        case color
        case weightCurrent = "weight_current"
        case isNeutered = "is_neutered"
        case allergies
        case specialNeeds = "special_needs"
    }

    var speciesSymbol: String {
        switch species.lowercased() {
        case "cat": return "cat"
        case "dog": return "dog"
        case "bird": return "bird"
        case "fish": return "fish"
        default: return "pawprint.fill"
        }
    }

    var ageString: String? {
        guard let raw = dateOfBirth else { return nil }
        let iso = DateFormatter()
        iso.dateFormat = "yyyy-MM-dd"
        iso.locale = Locale(identifier: "en_US_POSIX")
        let candidate = String(raw.prefix(10))
        guard let date = iso.date(from: candidate) else { return nil }
        let comps = Calendar.current.dateComponents([.year, .month], from: date, to: .now)
        if let y = comps.year, y > 0 { return "\(y)y old" }
        if let m = comps.month, m > 0 { return "\(m)mo old" }
        return nil
    }

    var dobFormatted: String? {
        guard let raw = dateOfBirth else { return nil }
        let iso = DateFormatter()
        iso.dateFormat = "yyyy-MM-dd"
        iso.locale = Locale(identifier: "en_US_POSIX")
        guard let date = iso.date(from: String(raw.prefix(10))) else { return nil }
        let out = DateFormatter()
        out.dateStyle = .medium
        out.timeStyle = .none
        return out.string(from: date)
    }

    var weightDisplay: String? {
        guard let w = weightCurrent, let value = Double(w), value > 0 else { return nil }
        let formatted = value.truncatingRemainder(dividingBy: 1) == 0
            ? String(format: "%.0f kg", value)
            : String(format: "%.1f kg", value)
        return formatted
    }
}

// MARK: - Vaccinations

struct VaccinationsResponse: Decodable {
    let status: String
    let vaccinations: [Vaccination]
}

struct Vaccination: Codable, Identifiable {
    let vaccinationId: Int
    let vaccineName: String
    let vaccineType: String?
    let vaccinationDate: String
    let nextDueDate: String?
    let adverseReaction: Bool
    let notes: String?
    let administeredByName: String?

    var id: Int { vaccinationId }

    enum CodingKeys: String, CodingKey {
        case vaccinationId = "vaccination_id"
        case vaccineName = "vaccine_name"
        case vaccineType = "vaccine_type"
        case vaccinationDate = "vaccination_date"
        case nextDueDate = "next_due_date"
        case adverseReaction = "adverse_reaction"
        case notes
        case administeredByName = "administered_by_name"
    }

    private static let isoParser: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private static let displayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateStyle = .medium
        f.timeStyle = .none
        return f
    }()

    private func parseDate(_ raw: String) -> Date? {
        Self.isoParser.date(from: String(raw.prefix(10)))
    }

    var vaccinationDateFormatted: String {
        parseDate(vaccinationDate).map { Self.displayFormatter.string(from: $0) } ?? vaccinationDate
    }

    var nextDueDateFormatted: String? {
        guard let raw = nextDueDate else { return nil }
        return parseDate(raw).map { Self.displayFormatter.string(from: $0) } ?? raw
    }

    /// Returns nil if no due date, .past if overdue, .future if upcoming.
    enum DueStatus { case past, future }
    var dueStatus: DueStatus? {
        guard let raw = nextDueDate, let date = parseDate(raw) else { return nil }
        return date < Date() ? .past : .future
    }
}

// MARK: - Lab Reports

struct LabReportsResponse: Decodable {
    let status: String
    let reports: [LabReport]
}

struct LabReport: Codable, Identifiable {
    let reportId: Int
    let petId: String
    let reportName: String
    let reportType: String
    let fileType: String
    let notes: String?
    let createdAt: String

    var id: Int { reportId }

    enum CodingKeys: String, CodingKey {
        case reportId = "report_id"
        case petId = "pet_id"
        case reportName = "report_name"
        case reportType = "report_type"
        case fileType = "file_type"
        case notes
        case createdAt = "created_at"
    }

    var typeDisplayName: String {
        switch reportType {
        case "blood_test":   return "Blood Test"
        case "urinalysis":   return "Urinalysis"
        case "kidney_panel": return "Kidney Panel"
        case "x_ray":        return "X-Ray"
        case "ultrasound":   return "Ultrasound"
        case "cytology":     return "Cytology"
        case "biopsy":       return "Biopsy"
        case "culture":      return "Culture"
        default:             return "Other"
        }
    }

    var typeIcon: String {
        switch reportType {
        case "blood_test":   return "drop.fill"
        case "urinalysis":   return "flask.fill"
        case "kidney_panel": return "staroflife.fill"
        case "x_ray":        return "rays"
        case "ultrasound":   return "waveform.path.ecg"
        case "cytology":     return "magnifyingglass.circle.fill"
        case "biopsy":       return "stethoscope"
        case "culture":      return "testtube.2"
        default:             return "doc.text.fill"
        }
    }

    var formattedDate: String {
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let date = iso.date(from: createdAt) ?? ISO8601DateFormatter().date(from: createdAt)
        guard let d = date else { return String(createdAt.prefix(10)) }
        let fmt = DateFormatter()
        fmt.dateStyle = .medium
        fmt.timeStyle = .none
        return fmt.string(from: d)
    }
}

// MARK: - Change password

struct ChangePasswordFirstLoginRequest: Encodable {
    let newPassword: String
}

struct AckResponse: Decodable {
    let status: String
}

// MARK: - First-time account setup

struct VerifyIdentityRequest: Encodable {
    let email: String
    let phone: String
}

struct VerifyIdentityResponse: Decodable {
    let data: VerifyIdentityData
}

struct VerifyIdentityData: Decodable {
    let setupToken: String
    let firstName: String
}

struct SetFirstPasswordRequest: Encodable {
    let setupToken: String
    let newPassword: String
}

// MARK: - Appointments

struct Appointment: Codable, Identifiable {
    let appointmentId: String   // e.g. "APPT-2026-0001" (VARCHAR, trigger-generated)
    let petId: String
    let appointmentDate: String   // DATE from pg → "YYYY-MM-DDT00:00:00.000Z"
    let appointmentTime: String   // TIME from pg → "HH:MM:SS"
    let durationMinutes: Int
    let appointmentType: String
    let reason: String?
    let status: String
    let cancellationReason: String?
    let createdAt: String
    let veterinarianId: Int?
    let veterinarianName: String?
    let petName: String
    let species: String

    var id: String { appointmentId }

    enum CodingKeys: String, CodingKey {
        case appointmentId      = "appointment_id"
        case petId              = "pet_id"
        case appointmentDate    = "appointment_date"
        case appointmentTime    = "appointment_time"
        case durationMinutes    = "duration_minutes"
        case appointmentType    = "appointment_type"
        case reason
        case status
        case cancellationReason = "cancellation_reason"
        case createdAt          = "created_at"
        case veterinarianId     = "veterinarian_id"
        case veterinarianName   = "veterinarian_name"
        case petName            = "pet_name"
        case species
    }

    private static let dateParser: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(abbreviation: "UTC")
        return f
    }()

    var dateObject: Date? {
        Self.dateParser.date(from: String(appointmentDate.prefix(10)))
    }

    var dateFormatted: String {
        guard let d = dateObject else { return String(appointmentDate.prefix(10)) }
        let out = DateFormatter()
        out.dateStyle = .medium
        out.timeStyle = .none
        return out.string(from: d)
    }

    var timeFormatted: String {
        let raw = String(appointmentTime.prefix(5))
        let f = DateFormatter()
        f.dateFormat = "HH:mm"
        f.locale = Locale(identifier: "en_US_POSIX")
        guard let d = f.date(from: raw) else { return raw }
        let out = DateFormatter()
        out.timeStyle = .short
        out.dateStyle = .none
        return out.string(from: d)
    }

    var typeDisplayName: String {
        switch appointmentType {
        case "checkup":      return "Check-up"
        case "vaccination":  return "Vaccination"
        case "follow_up":    return "Follow-up"
        case "consultation": return "Consultation"
        default:             return appointmentType.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }

    var typeIcon: String {
        switch appointmentType {
        case "checkup":      return "stethoscope"
        case "vaccination":  return "syringe"
        case "follow_up":    return "arrow.clockwise"
        case "consultation": return "text.bubble"
        default:             return "calendar"
        }
    }

    var statusDisplayName: String {
        switch status {
        case "confirmed":   return "Confirmed"
        case "scheduled":   return "Scheduled"
        case "in_progress": return "In Progress"
        case "completed":   return "Completed"
        case "cancelled":   return "Cancelled"
        case "no_show":     return "No Show"
        default:            return status.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }

    var isActiveStatus: Bool {
        status == "confirmed" || status == "scheduled"
    }

    var isUpcoming: Bool {
        isActiveStatus || status == "in_progress"
    }

    // Returns true when the appointment is at least 48 hours away and still modifiable.
    var canModify: Bool {
        guard isActiveStatus else { return false }
        let datePart = String(appointmentDate.prefix(10))
        let timePart = String(appointmentTime.prefix(5))
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd HH:mm"
        guard let target = f.date(from: "\(datePart) \(timePart)") else { return false }
        return target.timeIntervalSinceNow >= 48 * 3600
    }
}

struct AppointmentsResponse: Decodable {
    let status: String
    let data: AppointmentsData
}

struct AppointmentsData: Decodable {
    let appointments: [Appointment]
}

struct AppointmentSingleResponse: Decodable {
    let status: String
    let message: String?
    let data: AppointmentSingleData
}

struct AppointmentSingleData: Decodable {
    let appointment: Appointment
}

// MARK: - Veterinarians list

struct Veterinarian: Codable, Identifiable {
    let veterinarianId: Int
    let firstName: String
    let lastName: String
    let specialization: String?

    var id: Int { veterinarianId }
    var fullName: String { "\(firstName) \(lastName)" }

    enum CodingKeys: String, CodingKey {
        case veterinarianId = "veterinarian_id"
        case firstName      = "first_name"
        case lastName       = "last_name"
        case specialization
    }
}

struct VeterinariansResponse: Decodable {
    let status: String
    let data: VeterinariansData
}

struct VeterinariansData: Decodable {
    let veterinarians: [Veterinarian]
}

// MARK: - Availability

struct AvailabilityResponse: Decodable {
    let status: String
    let data: AvailabilityData
}

struct AvailabilityData: Decodable {
    let date: String
    let isClinicDay: Bool
    let slots: [TimeSlot]
}

struct TimeSlot: Codable, Identifiable {
    let time: String
    let remainingCapacity: Int
    let available: Bool
    let meetsLeadTime: Bool

    var id: String { time }

    var timeFormatted: String {
        let f = DateFormatter()
        f.dateFormat = "HH:mm"
        f.locale = Locale(identifier: "en_US_POSIX")
        guard let d = f.date(from: time) else { return time }
        let out = DateFormatter()
        out.timeStyle = .short
        out.dateStyle = .none
        return out.string(from: d)
    }
}

// MARK: - Appointment booking requests

struct CreateAppointmentRequest: Encodable {
    let petId: String
    let appointmentDate: String
    let appointmentTime: String
    let appointmentType: String
    let reason: String
    let veterinarianId: Int?

    enum CodingKeys: String, CodingKey {
        case petId           = "pet_id"
        case appointmentDate = "appointment_date"
        case appointmentTime = "appointment_time"
        case appointmentType = "appointment_type"
        case reason
        case veterinarianId  = "veterinarian_id"
    }
}

// MARK: - Profile update

struct UpdateProfileRequest: Encodable {
    let alternatePhone: String?
    let address: String?
    let city: String?
    let preferredContactMethod: String?
    let emergencyContact: String?
    let emergencyPhone: String?

    enum CodingKeys: String, CodingKey {
        case alternatePhone         = "alternate_phone"
        case address
        case city
        case preferredContactMethod = "preferred_contact_method"
        case emergencyContact       = "emergency_contact"
        case emergencyPhone         = "emergency_phone"
    }
}

struct CustomerUpdateResponse: Decodable {
    let status: String
    let data: CustomerUpdateData
}

struct CustomerUpdateData: Decodable {
    let customer: Customer
}

struct UpdateAppointmentRequest: Encodable {
    let appointmentDate: String
    let appointmentTime: String
    let appointmentType: String
    let reason: String
    let veterinarianId: Int?

    enum CodingKeys: String, CodingKey {
        case appointmentDate = "appointment_date"
        case appointmentTime = "appointment_time"
        case appointmentType = "appointment_type"
        case reason
        case veterinarianId  = "veterinarian_id"
    }
}

File: mobile/android/app/src/main/java/com/vetcare/pro/data/model/CustomerModels.kt
package com.vetcare.pro.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

private fun String.capitalizeFirst(): String {
    return this.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.US) else it.toString() }
}

@Serializable
data class CustomerLoginRequest(
    val identifier: String,
    val password: String
)

@Serializable
data class CustomerLoginResponse(
    val status: String? = null,
    val message: String? = null,
    val data: CustomerLoginData
)

@Serializable
data class CustomerLoginData(
    val customer: Customer,
    val token: String
)

@Serializable
data class Customer(
    @SerialName("customer_id") val customerId: String,
    @SerialName("first_name") val firstName: String,
    @SerialName("last_name") val lastName: String,
    val email: String? = null,
    val phone: String,
    @SerialName("alternate_phone") val alternatePhone: String? = null,
    val address: String? = null,
    val city: String? = null,
    val nic: String? = null,
    @SerialName("preferred_contact_method") val preferredContactMethod: String? = null,
    @SerialName("emergency_contact") val emergencyContact: String? = null,
    @SerialName("emergency_phone") val emergencyPhone: String? = null,
    @SerialName("password_must_change") val passwordMustChange: Boolean = false,
    @SerialName("created_at") val createdAt: String? = null
) {
    val fullName: String get() = "$firstName $lastName"

    val initials: String
        get() = "${firstName.take(1)}${lastName.take(1)}".uppercase()

    val memberSince: String?
        get() {
            val raw = createdAt ?: return null
            return try {
                val parser = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val date = parser.parse(raw.take(10))
                val cal = java.util.Calendar.getInstance()
                if (date != null) cal.time = date
                cal.get(java.util.Calendar.YEAR).toString()
            } catch (e: Exception) {
                null
            }
        }
}

@Serializable
data class PetsResponse(
    val status: String? = null,
    val data: PetsData
)

@Serializable
data class PetsData(
    val pets: List<Pet>
)

@Serializable
data class Pet(
    @SerialName("pet_id") val petId: String,
    @SerialName("pet_name") val name: String,
    val species: String,
    val breed: String? = null,
    val gender: String? = null,
    @SerialName("date_of_birth") val dateOfBirth: String? = null,
    val color: String? = null,
    @SerialName("weight_current") val weightCurrent: String? = null,
    @SerialName("is_neutered") val isNeutered: Boolean = false,
    val allergies: String? = null,
    @SerialName("special_needs") val specialNeeds: String? = null
) {
    val ageString: String?
        get() {
            val raw = dateOfBirth ?: return null
            return try {
                val fmt = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val date = fmt.parse(raw.take(10)) ?: return null
                val now = Date()
                val diffMs = now.time - date.time
                val days = diffMs / (1000 * 60 * 60 * 24)
                val years = days / 365
                if (years > 0) "$years y old"
                else {
                    val months = days / 30
                    if (months > 0) "$months mo old" else null
                }
            } catch (e: Exception) {
                null
            }
        }

    val dobFormatted: String?
        get() {
            val raw = dateOfBirth ?: return null
            return try {
                val parser = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val date = parser.parse(raw.take(10)) ?: return null
                val out = SimpleDateFormat("MMM d, yyyy", Locale.US)
                out.format(date)
            } catch (e: Exception) {
                null
            }
        }

    val weightDisplay: String?
        get() {
            val w = weightCurrent?.toDoubleOrNull() ?: return null
            if (w <= 0) return null
            return if (w % 1.0 == 0.0) String.format(Locale.US, "%.0f kg", w)
            else String.format(Locale.US, "%.1f kg", w)
        }
}

@Serializable
data class VaccinationsResponse(
    val status: String,
    val vaccinations: List<Vaccination>
)

@Serializable
data class Vaccination(
    @SerialName("vaccination_id") val vaccinationId: Int,
    @SerialName("vaccine_name") val vaccineName: String,
    @SerialName("vaccine_type") val vaccineType: String? = null,
    @SerialName("vaccination_date") val vaccinationDate: String,
    @SerialName("next_due_date") val nextDueDate: String? = null,
    @SerialName("adverse_reaction") val adverseReaction: Boolean = false,
    val notes: String? = null,
    @SerialName("administered_by_name") val administeredByName: String? = null
) {
    enum class DueStatus { OVERDUE, UPCOMING }

    val dueStatus: DueStatus?
        get() {
            val raw = nextDueDate ?: return null
            return try {
                val fmt = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val dueDate = fmt.parse(raw.take(10)) ?: return null
                if (dueDate.before(Date())) DueStatus.OVERDUE else DueStatus.UPCOMING
            } catch (e: Exception) {
                null
            }
        }

    val vaccinationDateFormatted: String
        get() {
            return try {
                val fmt = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val date = fmt.parse(vaccinationDate.take(10)) ?: return vaccinationDate
                val out = SimpleDateFormat("MMM d, yyyy", Locale.US)
                out.format(date)
            } catch (e: Exception) {
                vaccinationDate
            }
        }

    val nextDueDateFormatted: String?
        get() {
            val raw = nextDueDate ?: return null
            return try {
                val fmt = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val date = fmt.parse(raw.take(10)) ?: return raw
                val out = SimpleDateFormat("MMM d, yyyy", Locale.US)
                out.format(date)
            } catch (e: Exception) {
                raw
            }
        }
}

@Serializable
data class LabReportsResponse(
    val status: String,
    val reports: List<LabReport>
)

@Serializable
data class LabReport(
    @SerialName("report_id") val reportId: Int,
    @SerialName("pet_id") val petId: String,
    @SerialName("report_name") val reportName: String,
    @SerialName("report_type") val reportType: String,
    @SerialName("file_type") val fileType: String,
    val notes: String? = null,
    @SerialName("created_at") val createdAt: String
) {
    val typeDisplayName: String
        get() = when (reportType) {
            "blood_test" -> "Blood Test"
            "urinalysis" -> "Urinalysis"
            "kidney_panel" -> "Kidney Panel"
            "x_ray" -> "X-Ray"
            "ultrasound" -> "Ultrasound"
            "cytology" -> "Cytology"
            "biopsy" -> "Biopsy"
            "culture" -> "Culture"
            else -> "Other"
        }

    val formattedDate: String
        get() {
            return try {
                val parser = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                val date = parser.parse(createdAt.take(10)) ?: return createdAt.take(10)
                val out = SimpleDateFormat("MMM d, yyyy", Locale.US)
                out.format(date)
            } catch (e: Exception) {
                createdAt.take(10)
            }
        }
}

@Serializable
data class ChangePasswordFirstLoginRequest(
    @SerialName("newPassword") val newPassword: String
)

@Serializable
data class ChangePasswordRequest(
    @SerialName("currentPassword") val currentPassword: String,
    @SerialName("newPassword") val newPassword: String
)

@Serializable
data class AckResponse(
    val status: String,
    val message: String? = null
)

@Serializable
data class VerifyIdentityRequest(
    val email: String,
    val phone: String
)

@Serializable
data class VerifyIdentityResponse(
    val status: String? = null,
    val data: VerifyIdentityData
)

@Serializable
data class VerifyIdentityData(
    @SerialName("setupToken") val setupToken: String,
    @SerialName("firstName") val firstName: String
)

@Serializable
data class SetFirstPasswordRequest(
    @SerialName("setupToken") val setupToken: String,
    @SerialName("newPassword") val newPassword: String
)

@Serializable
data class Appointment(
    @SerialName("appointment_id") val appointmentId: String,
    @SerialName("pet_id") val petId: String,
    @SerialName("appointment_date") val appointmentDate: String,
    @SerialName("appointment_time") val appointmentTime: String,
    @SerialName("duration_minutes") val durationMinutes: Int = 30,
    @SerialName("appointment_type") val appointmentType: String,
    val reason: String? = null,
    val status: String,
    @SerialName("cancellation_reason") val cancellationReason: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("veterinarian_id") val veterinarianId: Int? = null,
    @SerialName("veterinarian_name") val veterinarianName: String? = null,
    @SerialName("pet_name") val petName: String = "",
    val species: String = ""
) {
    val dateFormatted: String
        get() {
            return try {
                val parser = SimpleDateFormat("yyyy-MM-dd", Locale.US)
                parser.timeZone = TimeZone.getTimeZone("UTC")
                val date = parser.parse(appointmentDate.take(10)) ?: return appointmentDate.take(10)
                val out = SimpleDateFormat("MMM d, yyyy", Locale.US)
                out.format(date)
            } catch (e: Exception) {
                appointmentDate.take(10)
            }
        }

    val timeFormatted: String
        get() {
            return try {
                val parser = SimpleDateFormat("HH:mm", Locale.US)
                val date = parser.parse(appointmentTime.take(5)) ?: return appointmentTime.take(5)
                val out = SimpleDateFormat("h:mm a", Locale.US)
                out.format(date)
            } catch (e: Exception) {
                appointmentTime.take(5)
            }
        }

    val typeDisplayName: String
        get() = when (appointmentType) {
            "checkup" -> "Check-up"
            "vaccination" -> "Vaccination"
            "follow_up" -> "Follow-up"
            "consultation" -> "Consultation"
            else -> appointmentType.replace("_", " ").capitalizeFirst()
        }

    val statusDisplayName: String
        get() = when (status) {
            "confirmed" -> "Confirmed"
            "scheduled" -> "Scheduled"
            "in_progress" -> "In Progress"
            "completed" -> "Completed"
            "cancelled" -> "Cancelled"
            "no_show" -> "No Show"
            else -> status.replace("_", " ").capitalizeFirst()
        }

    val isActiveStatus: Boolean
        get() = status == "confirmed" || status == "scheduled"

    val isUpcoming: Boolean
        get() = isActiveStatus || status == "in_progress"
}

@Serializable
data class AppointmentsResponse(
    val status: String,
    val data: AppointmentsData
)

@Serializable
data class AppointmentsData(
    val appointments: List<Appointment>
)

@Serializable
data class AppointmentSingleResponse(
    val status: String,
    val message: String? = null,
    val data: AppointmentSingleData
)

@Serializable
data class AppointmentSingleData(
    val appointment: Appointment
)

@Serializable
data class Veterinarian(
    @SerialName("veterinarian_id") val veterinarianId: Int,
    @SerialName("first_name") val firstName: String,
    @SerialName("last_name") val lastName: String,
    val specialization: String? = null
) {
    val fullName: String get() = "Dr. $firstName $lastName"
}

@Serializable
data class VeterinariansResponse(
    val status: String,
    val data: VeterinariansData
)

@Serializable
data class VeterinariansData(
    val veterinarians: List<Veterinarian>
)

@Serializable
data class AvailabilityResponse(
    val status: String,
    val data: AvailabilityData
)

@Serializable
data class AvailabilityData(
    val date: String,
    @SerialName("isClinicDay") val isClinicDay: Boolean = true,
    val slots: List<TimeSlot>
)

@Serializable
data class TimeSlot(
    val time: String,
    @SerialName("remainingCapacity") val remainingCapacity: Int = 3,
    val available: Boolean = true,
    @SerialName("meetsLeadTime") val meetsLeadTime: Boolean = true
) {
    val timeFormatted: String
        get() {
            return try {
                val parser = SimpleDateFormat("HH:mm", Locale.US)
                val date = parser.parse(time.take(5)) ?: return time
                val out = SimpleDateFormat("h:mm a", Locale.US)
                out.format(date)
            } catch (e: Exception) {
                time
            }
        }
}

@Serializable
data class CreateAppointmentRequest(
    @SerialName("pet_id") val petId: String,
    @SerialName("appointment_date") val appointmentDate: String,
    @SerialName("appointment_time") val appointmentTime: String,
    @SerialName("appointment_type") val appointmentType: String,
    val reason: String,
    @SerialName("veterinarian_id") val veterinarianId: Int? = null
)

@Serializable
data class UpdateAppointmentRequest(
    @SerialName("appointment_date") val appointmentDate: String,
    @SerialName("appointment_time") val appointmentTime: String,
    @SerialName("appointment_type") val appointmentType: String,
    val reason: String,
    @SerialName("veterinarian_id") val veterinarianId: Int? = null
)

@Serializable
data class CancelAppointmentRequest(
    val reason: String? = null
)

@Serializable
data class UpdateProfileRequest(
    @SerialName("alternate_phone") val alternatePhone: String? = null,
    val address: String? = null,
    val city: String? = null,
    @SerialName("preferred_contact_method") val preferredContactMethod: String? = null,
    @SerialName("emergency_contact") val emergencyContact: String? = null,
    @SerialName("emergency_phone") val emergencyPhone: String? = null
)

@Serializable
data class CustomerUpdateResponse(
    val status: String,
    val data: CustomerUpdateData
)

@Serializable
data class CustomerUpdateData(
    val customer: Customer
)

@Serializable
data class PhotoGuidanceJob(
    @SerialName("job_id") val jobId: Int,
    @SerialName("pet_id") val petId: String,
    @SerialName("owner_note") val ownerNote: String? = null,
    val status: String,
    @SerialName("guidance_text") val guidanceText: String? = null,
    @SerialName("error_message") val errorMessage: String? = null,
    @SerialName("created_at") val createdAt: String? = null
) {
    val isFinished: Boolean get() = status == "completed" || status == "failed"

    val statusDisplay: String
        get() = when (status) {
            "pending" -> "Waiting to start…"
            "processing" -> "Analyzing photo…"
            "completed" -> "Ready"
            "failed" -> "Failed"
            else -> status.capitalizeFirst()
        }
}

@Serializable
data class PhotoGuidanceSubmitResponse(
    val job: PhotoGuidanceJob
)

@Serializable
data class PhotoGuidanceStatusResponse(
    val job: PhotoGuidanceJob
)

@Serializable
data class PhotoGuidanceHistoryResponse(
    val jobs: List<PhotoGuidanceJob>
)

